internal/pkg/gateway/api.go

```diff
@@ -427,11 +428,81 @@ func (gs *Server) Submit(ctx context.Context, request *gp.SubmitRequest) (*gp.Su
 		return nil, status.Errorf(codes.Unavailable, "no orderer nodes available")
 	}
 
+	logger := logger.With("txID", request.TransactionId)
+	config := gs.getChannelConfig(request.ChannelId)
+	if config.ChannelConfig().Capabilities().ConsensusTypeBFT() {
+		return gs.submitBFT(ctx, orderers, txn, config, logger)
+	} else {
+		return gs.submitNonBFT(ctx, orderers, txn, logger)
+	}
+}
+
+func (gs *Server) submitBFT(ctx context.Context, orderers []*orderer, txn *common.Envelope, config channelconfig.Resources, logger *flogging.FabricLogger) (*gp.SubmitResponse, error) {
+	// For BFT, we send transaction to ALL orderers
+	ctx, cancel := context.WithCancel(ctx)
```

Comment:
The problem now is that you may finish sending to the quorum before you send to the leader, and the function will return, cancelling the context which will cancel sending to the leader. 

While the transaction will be scheduled eventually, it will be only after a timeout, and this is bad for user experience. 

Let's do something like this:


```
diff --git a/internal/pkg/gateway/api.go b/internal/pkg/gateway/api.go
index d2eebcabe..665fac1cb 100644
--- a/internal/pkg/gateway/api.go
+++ b/internal/pkg/gateway/api.go
@@ -12,6 +12,8 @@ import (
        "io"
        "math/rand"
        "strings"
+       "sync/atomic"
+       "time"
 
        "github.com/golang/protobuf/proto"
        "github.com/hyperledger/fabric-protos-go/common"
@@ -438,6 +440,10 @@ func (gs *Server) Submit(ctx context.Context, request *gp.SubmitRequest) (*gp.Su
 }
 
 func (gs *Server) submitBFT(ctx context.Context, orderers []*orderer, txn *common.Envelope, config channelconfig.Resources, logger *flogging.FabricLogger) (*gp.SubmitResponse, error) {
+       var numFinishedSend uint32
+
+       earliestFinishDeadline := time.Now().Add(time.Millisecond * 500) // TODO: Probably should be configurable
+
        // For BFT, we send transaction to ALL orderers
        ctx, cancel := context.WithCancel(ctx)
        defer cancel()
@@ -448,6 +454,7 @@ func (gs *Server) submitBFT(ctx context.Context, orderers []*orderer, txn *commo
                        ctx, cancel := context.WithTimeout(ctx, gs.options.BroadcastTimeout)
                        defer cancel()
                        response, err := gs.broadcast(ctx, ord, txn)
+                       atomic.AddUint32(&numFinishedSend, 1)
 
                        if err != nil {
                                logger.Warnw("Error sending transaction to orderer", "endpoint", ord.logAddress, "err", err)
@@ -483,6 +490,9 @@ loop:
                        } else {
                                successes++
                                if successes >= quorum {
+                                       if atomic.LoadUint32(&numFinishedSend) < uint32(n) {
+                                               time.Sleep(time.Until(earliestFinishDeadline))
+                                       }
                                        return &gp.SubmitResponse{}, nil
                                }
                        }

```

This ensures we will always linger enough time for the leader to receive the transaction if it is not overloaded and not faulty. 

We probably should export this to a config parameter.